keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+...
+...
+	// Recover all images.
+    <mark>
+	listImagesOpt := []containerd.ListImageOpt {
+		containerd.ListImageWithPlatformMatcherMap(c.platformMatcherMap),
+	}
+	cImages, err := c.client.ListImagesWithPlatformMatcher(ctx, listImagesOpt...) </mark>
+	if err != nil {
+		return fmt.Errorf("failed to list images: %w", err)
+	}
+	c.loadImages(ctx, cImages)
+...
+...
+}
+
+
+<mark>
+// New functions added in client.go:
+</mark>
+
+func checkIfValidPlatforMatcher(ctx context.Context, provider content.Provider, image ocispec.Descriptor, matcher platforms.MatchComparer) bool {
+	_, err := images.Manifest(ctx, provider, image, matcher)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+type ImagesWrapper struct {
+	Images Image
+	RuntimeHandler string
+}
+
+func (c *Client) ListImagesWithPlatformMatcher(ctx context.Context, opts ...ListImageOpt) ([]ImagesWrapper, error) {
+	var listImgOpts ListImageOptions
+	for _, o := range opts {
+		if err := o(ctx, &listImgOpts); err != nil {
+			return nil, err
+		}
+	}
+
+	imgs, err := c.ImageService().List(ctx, listImgOpts.Filters...)
+	if err != nil {
+		return nil, err
+	}
+
+	imagesW := make([]ImagesWrapper, len(imgs))
+	if len(opts) == 0 { 
+		for i, img := range imgs {
+			imagesW[i].Images = NewImage(c, img)
+			imagesW[i].RuntimeHandler = c.runtime
+		}
+	} else {
+		for i, img := range imgs {
+			for runtimeHandler, matcher := range listImgOpts.PlatformMatcherMap {
+				tmp := NewImage(c, img)
+				if checkIfValidPlatforMatcher(ctx, tmp.ContentStore(), tmp.Target(), matcher) {
+					imagesW[i].Images = NewImageWithPlatform(c, img, matcher)
+					imagesW[i].RuntimeHandler = runtimeHandler
+				}
+			}
+			if imagesW[i].Images == nil {
+				return nil, fmt.Errorf("valid platform matcher for image %v not found", img)
+			}
+		}
+	}
+	return imagesW, nil
+}
+
+</pre>
+
+
+### Windows platform matcher optional changes
+
+Windows platform matcher could optionally be extended to also check for host and UVM compatibility if hyperV isolation is being used based on the matrix described here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
+
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtime class:
+
+1. Before running a container, kuebelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
+
+PullImageRequest has a field for ImageSpec which can hold info for the runtime class used to pull this image:
```

Comment:
If nothing is filled in, the ImagePullRequest will have an empty field for runtime class and containerd would check this and use default runtime handler for the platform if none was specified. It would need changes in pkg/kubelet/images/puller.go pullImage() function  and EnsureImageExists() function in pkg/kubelet/images/image_manager.go. They do have access to the ImageSpec and the runtime handlers need to be set by the called. In short, there are changes that will be needed.